src/engine/network/scripting.game.ts

```diff
@@ -0,0 +1,100 @@
+import { createCursorView, readArrayBuffer, sliceCursorView, writeArrayBuffer } from "../allocator/CursorView";
+import { GameState } from "../GameTypes";
+import { defineModule, getModule } from "../module/module.common";
+import { ScriptResourceManager } from "../resource/ScriptResourceManager";
+import { NetworkModule } from "./network.game";
+import { NetworkAction } from "./NetworkAction";
+import { broadcastReliable } from "./outbound.game";
+import { NetPipeData, writeMetadata } from "./serialization.game";
+import { writeUint32, readUint32 } from "../allocator/CursorView";
+import { registerInboundMessageHandler } from "./inbound.game";
+
+interface WebSGNetworkModuleState {
+  scriptToInbound: Map<number, ArrayBuffer[]>;
+}
+
+export const WebSGNetworkModule = defineModule<GameState, WebSGNetworkModuleState>({
+  name: "WebSGNetwork",
+  create: () => {
+    return {
+      scriptToInbound: new Map(),
+    };
+  },
+  init(ctx: GameState) {
+    const network = getModule(ctx, NetworkModule);
+    registerInboundMessageHandler(network, NetworkAction.ScriptMessage, deserializeScriptMessage);
+  },
+});
+
+export function createWebSGNetworkModule(ctx: GameState, resourceManager: ScriptResourceManager) {
+  const network = getModule(ctx, NetworkModule);
+  const { scriptToInbound } = getModule(ctx, WebSGNetworkModule);
+  const { U8Heap } = resourceManager;
+
+  return {
+    network_broadcast: (packetPtr: number, byteLength: number) => {
+      try {
+        const scriptPacket = U8Heap.subarray(packetPtr, packetPtr + byteLength);
+
+        const msg = createScriptMessage(ctx, resourceManager.id!, scriptPacket);
+
+        broadcastReliable(ctx, network, msg);
+
+        return 0;
+      } catch (error) {
+        console.error("Error broadcasting packet:", error);
+        return -1;
+      }
+    },
+    // call until 0 return
+    network_receive: (writeBufPtr: number, maxBufLength: number) => {
+      const inbound = scriptToInbound.get(resourceManager.id!);
+      const packet = inbound!.pop();
+
+      if (!packet) {
+        return 0;
+      }
+
+      if (packet.byteLength > maxBufLength) {
+        console.error("Failed to receive script packet, packet length exceeded buffer length");
+        return -1;
+      }
+
+      try {
+        U8Heap.set(new Uint8Array(packet), writeBufPtr);
+        return 1;
```

Comment:
I think we should return the actual received packet length instead of just 1 here. That lets us allocate a fixed size buffer to read into but have different packet sizes being sent over the wire.